Add optimized quantize function for ARM

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26867 Add optimized quantize function for ARM**

Summary:
Use caffe2::Int8Quantize for pytorch mobile. Currently this is only implemented for uint8 tensors and runs using NEON intrinsics.
For all other cases it falls back to naive pytorch quantize_val implementation.

Previously, naive implementation of quantize_val is slow on mobile, taking up more than 50% of the execution time.

Results
Before
aten::quantize_per_tensor 42.893 ms
Total model runtime 70.5ms

After
aten::quantize_per_tensor 0.340 ms
Total model runtime 27.5ms

Test Plan:
Tested current python tests work python test/test_quantized.py TestQNNPackOps
Also tested using quantized mobilenetV2 on mobile and compared output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17638732](https://our.internmc.facebook.com/intern/diff/D17638732)

supriyar committed:
Add optimized quantize function for ARM
Summary:
Use caffe2::Int8Quantize for pytorch mobile. Currently this is only implemented for uint8 tensors and runs using NEON intrinsics.
For all other cases it falls back to naive pytorch quantize_val implementation.

Previously, naive implementation of quantize_val is slow on mobile, taking up more than 50% of the execution time.

Results
Before
aten::quantize_per_tensor 42.893 ms
Total model runtime 70.5ms

After
aten::quantize_per_tensor 0.340 ms
Total model runtime 27.5ms

Test Plan:
Tested current python tests work python test/test_quantized.py TestQNNPackOps
Also tested using quantized mobilenetV2 on mobile and compared output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add optimized quantize function for ARM"


Summary:
Use caffe2::Int8Quantize for pytorch mobile. Currently this is only implemented for uint8 tensors and runs using NEON intrinsics.
For all other cases it falls back to naive pytorch quantize_val implementation.

Previously, naive implementation of quantize_val is slow on mobile, taking up more than 50% of the execution time.

Results
Before
aten::quantize_per_tensor 42.893 ms
Total model runtime 70.5ms

After
aten::quantize_per_tensor 0.340 ms
Total model runtime 27.5ms

Test Plan:
Tested current python tests work python test/test_quantized.py TestQNNPackOps
Also tested using quantized mobilenetV2 on mobile and compared output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+inline uint8_t Uint8Quantize(const float scale, const int32_t zero_point, const float value) {
+  const int32_t qmin = std::numeric_limits<uint8_t>::min();
+  const int32_t qmax = std::numeric_limits<uint8_t>::max();
+  auto r = zero_point + static_cast<int32_t>(Round(value / scale));
```
Couple of questions:
1. Does this match what we do in FBGEMM? i.e adding zero-point after rounding:
2. We should add a test to ensure that Quantize and DeQuantize are identical across backends. I know they mismatched before, but lets change this to come up with a bit-exact match.

supriyar replied:
No this is a difference right now. FBGEMM does `f = static_cast<int64_t>(Round(value / scale + zero_point));`
caffe2 Int8Quantize does `f= zero_point + static_cast<int32_t>(Round(value / scale));`

I've added it as a TODO to update the pt mobile implementation to be on par with fbgemm.

raghuramank100 commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+  }
+#endif
+  for (; i < N; ++i) {
+    (*out++) = Uint8Quantize(Y_scale, Y_offset, (*in++));
```
Confused by the flow here: Int8Quantize calls Uint8Quantize inside, which is confusing. Also, if Uint8Quantize does the rounding internally, whats the purpose of the simd code above?

supriyar replied:
This is for the leftover elements. simd code above works for input elements using stride of 8.

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
   }
 }
 
+#ifdef GEMMLOWP_NEON
```
looks like the only header you need is third_party/gemmlowp/gemmlowp/internal/detect_platform.h . And maybe even that better to pull out if we don't depend on gemmlowp otherwise

supriyar replied:
Removed this and add an explicit `__ARM_NEON__` define to guard this.

raghuramank100 commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+  const bool use_pytorch_qnnpack_arm = false;
+#endif
+
+  // Int8Quantize only works with uint8 type.
```
Why is this the case? Can we rename it to Uint8Quantize?

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+#define INT8_NEON_SIMD
+#endif
+
+#if defined(__SSE4_2__) && defined(__clang__)
```
keep the original comment in this code. Also - this logic might be nice to pull out to a separate header file

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
-    qdata[i] = quantize_val<T>(scale, zero_point, rdata[i]);
+  // If PYTORCH_QNNPACK is enabled, use caffe2 specialized Int8Quantize implementation
+#if defined(USE_PYTORCH_QNNPACK) && defined(INT8_NEON_SIMD)
+  const bool use_pytorch_qnnpack_arm = at::globalContext().qEngine() == at::QEngine::QNNPACK;
```
I think it's sufficient to just check the engine - it'd be set only if we're built with qnnpack. Or if we're confident in this logic and don't want to run it using clang on server side - we should just use magical implementation iff the neon is present

supriyar replied:
I run the logic now explicitly if neon is present. However the code only works for uint8, so for all other cases it defaults to the default implementation.

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+  // same number (0x4B400000 is the integer representation of 12582912.0f) to
+  // get only the mantissa. This works if -2**22 < x < 2**22, but preserves the
+  // sign for negative numbers.
+  const int32x4_t voffset = vdupq_n_s32(Y_offset - 0x4B400000);
```
even though I'm fine landing it like this for demo, I think we should fix it to have the same offset addition logic as server side. Can you add a TODO?

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+#include "NEON_2_SSE.h"
+#define INT8_NEON_SIMD
+#endif
+inline uint8_t Uint8Quantize(const float scale, const int32_t zero_point, const float value) {
```
this should really be merged with quantize_val above after we move in the zero point inside the round (fine as TODO + separate PR)

supriyar replied:
Added a TODO

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +187,109 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+}
+
 template <typename T>
 Tensor quantize_tensor(Tensor rtensor, Tensor qtensor, double scale, int64_t zero_point) {
```
btw, this function might benefit from TORCH_CHECK(rtensor.is_contiguous())

dzhulgakov reviewed (changes requested):
nice!

supriyar committed:
Update on "Add optimized quantize function for ARM"


Summary:
Use caffe2::Int8Quantize for pytorch mobile. Currently this is only implemented for uint8 tensors and runs using NEON intrinsics.
For all other cases it falls back to naive pytorch quantize_val implementation.

Previously, naive implementation of quantize_val is slow on mobile, taking up more than 50% of the execution time.

Results
Before
aten::quantize_per_tensor 42.893 ms
Total model runtime 70.5ms

After
aten::quantize_per_tensor 0.340 ms
Total model runtime 27.5ms

Test Plan:
Tested current python tests work python test/test_quantized.py TestQNNPackOps
Also tested using quantized mobilenetV2 on mobile and compared output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +189,121 @@ T quantize_val(double scale, int64_t zero_point, float value) {
   int64_t qvalue;
   constexpr int64_t qmin = std::numeric_limits<typename T::underlying>::min();
   constexpr int64_t qmax = std::numeric_limits<typename T::underlying>::max();
-  checkZeroPoint<typename T::underlying>("quantize_val", zero_point);
```
why is this removed?

supriyar replied:
We were calling this for each value, moved it out of the for loop.
Saves about 20ms on mobile :)

jerryzh168 commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +189,121 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+
+// Generic template defaults to naive quantize implementation
+template <typename T>
+void Int8Quantize(
```
why don't we use the same function interface as `quantize_tensor`?

jerryzh168 replied:
e.g. `quantize_tensor_arm`

jerryzh168 commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +189,121 @@ T quantize_val(double scale, int64_t zero_point, float value) {
-  const float* rdata = rtensor.data_ptr<float>();
+  TORCH_CHECK(rtensor.is_contiguous(), "Float tensor should be contiguous");
+  const float* const rdata = rtensor.data_ptr<float>();
   auto qdata = qtensor.data_ptr<T>();
```
qdata seems to be only used in the `else` branch, maybe move this there?

jerryzh168 commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +189,121 @@ T quantize_val(double scale, int64_t zero_point, float value) {
   checkZeroPoint<typename T::underlying>(fn_name, zero_point);
-  const float* rdata = rtensor.data_ptr<float>();
+  TORCH_CHECK(rtensor.is_contiguous(), "Float tensor should be contiguous");
+  const float* const rdata = rtensor.data_ptr<float>();
```
this can be removed if we change the function signature of Int8Quantize.

supriyar replied:
We will still need it as quantize_val below requires it

supriyar committed:
Update on "Add optimized quantize function for ARM"


Summary:
Use caffe2::Int8Quantize for pytorch mobile. Currently this is only implemented for uint8 tensors and runs using NEON intrinsics.
For all other cases it falls back to naive pytorch quantize_val implementation.

Previously, naive implementation of quantize_val is slow on mobile, taking up more than 50% of the execution time.

Results
Before
aten::quantize_per_tensor 42.893 ms
Total model runtime 70.5ms

After
aten::quantize_per_tensor 0.340 ms
Total model runtime 27.5ms

Test Plan:
Tested current python tests work python test/test_quantized.py TestQNNPackOps
Also tested using quantized mobilenetV2 on mobile and compared output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

resistor commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -159,6 +162,21 @@ Tensor dequantize_tensor(Tensor qtensor, Tensor rtensor, double scale, int64_t z
+#if defined(__ANDROID__) && !defined(__NDK_MAJOR__)
+template <class T>
+inline float Round(const float x) {
+  return ::nearbyintf(x);
```
Are you sure you want `nearbyint` rather than `round`?  The latter ignores rounding mode flags and is usually more performant.

supriyar replied:
Thanks for pointing it out, I was trying to be more consistent with what we're already using today in caffe2 and fbgemm::Quantize (https://github.com/pytorch/FBGEMM/blob/7f55d6c14fb8ff2b0b03ddf9c4166bd052460fec/include/fbgemm/QuantUtils.h#L65)

resistor commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,30 +189,122 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+    vst1_u8(out, vout01234567);
+    out += 8;
+  }
+#endif // __ARM_NEON__
```
Shouldn't there be a `#else`?  I don't see how this code can be correct for both NEON and non-NEON as written.  At the least I'd expect to see a `static_assert` in the `#else`.

supriyar replied:
For non-NEON architectures, it will fall back to the naive implementation in the for loop below.

resistor replied:
Can you move the `#ifdef` around the entire function then?  In the current version on non-ARM platforms we will be generating compiling and generating code for this function that does not implement it.  I'd rather it not be present rather than present but non-working, both to save code size and to reduce the chances of someone calling it improperly in the future.

supriyar committed:
Update on "Add optimized quantize function for ARM"


Summary:
Use caffe2::Int8Quantize for pytorch mobile. Currently this is only implemented for uint8 tensors and runs using NEON intrinsics.
For all other cases it falls back to naive pytorch quantize_val implementation.

Previously, naive implementation of quantize_val is slow on mobile, taking up more than 50% of the execution time.

Results
Before
aten::quantize_per_tensor 42.893 ms
Total model runtime 70.5ms

After
aten::quantize_per_tensor 0.340 ms
Total model runtime 27.5ms

Test Plan:
Tested current python tests work python test/test_quantized.py TestQNNPackOps
Also tested using quantized mobilenetV2 on mobile and compared output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@pytorchbot retest this please

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -6,11 +6,14 @@
 
 #ifdef USE_FBGEMM
 #include <fbgemm/QuantUtils.h>
+#else
```
FBGEMM won't be on ARM - so you can remove #else, it'd make it more readable

dzhulgakov commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -171,32 +189,120 @@ T quantize_val(double scale, int64_t zero_point, float value) {
+    quantize_tensor_arm<T>(rdata, qtensor, rtensor.numel(), scale, zero_point);
+    return qtensor;
+  }
+#else
```
I think you shouldn't have this #else or you should just ignore qengine. With current logic if someone sets qengine to none on arm - the entire function won't do anything!

supriyar replied:
Ideally this function should only be called when we run a quantized model on ARM, and if that is the case them we currently only support that if qengine is set to qnnpack. But I'm fine with removing it as it will fail later in the op anyway.

supriyar committed:
Update on "Add optimized quantize function for ARM"


Summary:
Use caffe2::Int8Quantize for pytorch mobile. Currently this is only implemented for uint8 tensors and runs using NEON intrinsics.
For all other cases it falls back to naive pytorch quantize_val implementation.

Previously, naive implementation of quantize_val is slow on mobile, taking up more than 50% of the execution time.

Results
Before
aten::quantize_per_tensor 42.893 ms
Total model runtime 70.5ms

After
aten::quantize_per_tensor 0.340 ms
Total model runtime 27.5ms

Test Plan:
Tested current python tests work python test/test_quantized.py TestQNNPackOps
Also tested using quantized mobilenetV2 on mobile and compared output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@pytorchbot retest this please
